src/infill/LightningTree.cpp

```diff
@@ -0,0 +1,384 @@
+    //       (NOTE: in case realign turns out _not_ to be last, would need to rewrite a few things, see the 'rerooted_parts' parameter of that function).
+    tree_below->prune(prune_distance);
+    tree_below->straighten(smooth_magnitude);
+    if (tree_below->realign(next_outlines, next_trees))
+    {
+        next_trees.push_back(tree_below);
+    }
+}
+
+// NOTE: Depth-first, as currently implemented.
+//       Skips the root (because that has no root itself), but all initial nodes will have the root point anyway.
+void LightningTreeNode::visitBranches(const std::function<void(const Point&, const Point&)>& visitor) const
+{
+    for (const auto& node : children)
+    {
+        assert(node->parent.lock() == shared_from_this());
+        visitor(p, node->p);
+        node->visitBranches(visitor);
+    }
+}
+
+// NOTE: Depth-first, as currently implemented.
+void LightningTreeNode::visitNodes(const std::function<void(std::shared_ptr<LightningTreeNode>)>& visitor)
+{
+    visitor(shared_from_this());
+    for (const auto& node : children)
+    {
+        assert(node->parent.lock() == shared_from_this());
+        node->visitNodes(visitor);
+    }
+}
+
+LightningTreeNode::LightningTreeNode(const Point& p)
+: is_root(true)
+, p(p)
+{}
+
+void LightningTreeNode::findClosestNodeHelper(const Point& x, const coord_t supporting_radius, coord_t& closest_distance, std::shared_ptr<LightningTreeNode>& closest_node)
+{
+    for (const auto& node : children)
+    {
+        node->findClosestNodeHelper(x, supporting_radius, closest_distance, closest_node);
+        const coord_t distance = node->getWeightedDistance(x, supporting_radius);
+        if (distance < closest_distance)
+        {
+            closest_node = node;
+            closest_distance = distance;
+        }
+    }
+}
+
+std::shared_ptr<LightningTreeNode> LightningTreeNode::deepCopy() const
+{
+    std::shared_ptr<LightningTreeNode> local_root = LightningTreeNode::create(p);
+    local_root->is_root = is_root;
+    local_root->children.reserve(children.size());
+    for (const auto& node : children)
+    {
+        std::shared_ptr<LightningTreeNode> child = node->deepCopy();
+        child->parent = local_root;
+        local_root->children.push_back(child);
+    }
+    return local_root;
+}
+
+bool LightningTreeNode::realign(const Polygons& outlines, std::vector<std::shared_ptr<LightningTreeNode>>& rerooted_parts, const bool& connected_to_parent)
+{
+    // TODO: Hole(s) in the _middle_ of a line-segement, not unlikely since reconnect.
+    // TODO: Reconnect if not on outline -> plan is: yes, but not here anymore!
+
+    if (outlines.empty())
+    {
+        return false;
+    }
+
+    if (outlines.inside(p, true))
+    {
+        // Only keep children that have an unbroken connection to here, realign will put the rest in rerooted parts due to recursion:
+        const std::function<bool(const std::shared_ptr<LightningTreeNode>& child)> remove_unconnected_func
```

Comment:
I could argue that the return type is immediately clear to the programmer... but it's fine in this case, since we use it in the erase, which should give enough conext.